File: HW1/task3.py
from pyspark import SparkContext
import json
import sys
import time


if __name__ == '__main__':
    time1 = time.time()
    sc = SparkContext(master="local[*]", appName="task3")
    sc.setLogLevel("ERROR")
    input_file = sys.argv[1]
    output = sys.argv[2]
    part_type = sys.argv[3]
    part = int(sys.argv[4])
    n = int(sys.argv[5])

    review_lines = sc.textFile(input_file)

    review_lines2 = review_lines.map(lambda x: (json.loads(x)['business_id'], 1))

    # After mapping, partition is lost because key is changed, so should partition again to increase speed
    if part_type == "customized":
        review_lines2 = review_lines2.partitionBy(part)

    # Count and filter for more than n reviews
    review_lines4 = review_lines2.reduceByKey(lambda x, y: x + y) \
        .filter(lambda x: x[1] > n)

    answer = dict()
    answer['n_partitions'] = review_lines2.getNumPartitions()
    answer['n_items'] = review_lines2.mapPartitions(lambda iter: [sum(1 for _ in iter)]).collect()
    num_reviews = review_lines4.collect()
    answer['result'] = [[str(elem[0]), elem[1]] for elem in num_reviews]

    # Write answers
    file = open(output, "w")
    file.write(json.dumps(answer))
    file.close()

    print("Final answers:")
    print(json.dumps(answer))
    print(time.time() - time1)
    exit()

File: HW1/task3.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._
import java.io.{File, FileWriter}
import scala.io.Source


object task3 {
  def main(args: Array[String]): Unit = {
    val ss = SparkSession.builder().appName("task3").config("spark.master", "local[*]").getOrCreate()
    val sc = ss.sparkContext
    sc.setLogLevel("ERROR")
    val input_file = args(0)
    val output = args(1)
    val part_type = args(2)
    val part = args(3).toInt
    val n = args(4).toInt

    val review_lines = sc.textFile(input_file)
    var review_lines2 = review_lines.map((x: String) => (compact(parse(x) \ "business_id"), 1))

    if (part_type == "customized") {
      review_lines2 = review_lines2.repartition(part)
    }
    // Count and filter for more than n reviews
    val review_lines4 = (review_lines2.reduceByKey((x, y) => x + y)
      .filter(x => x._2 >= n))

    val n_items: Array[Int] = review_lines2.mapPartitions(it => Array(it.size).iterator, true).collect()
    val n_partitions = review_lines2.partitions.length
    val num_reviews = review_lines4.collect()

    // Format results to Json4s format
    var result: List[JArray] = List()
    for (i <- num_reviews) {
      result :+= JArray(List(JString(i._1.replace("\"", "")), i._2))
    }
    val result1 = JArray(result)

    var n_items1: List[JInt] = List()
    for (i <- n_items) {
      n_items1 :+= JInt(i)
    }
    val n_items2 = JArray(n_items1)

    val answer: JObject = (("n_partitions", n_partitions) ~ ("n_items", n_items2) ~ ("result", result1))

    val file = new File(output)
    val fw = new FileWriter(file)
    fw.write(compact(answer))
    fw.close()
  }
}
